### ISSUE TITLE
uri module doesn't properly handle escape sequences in JSON

### ISSUE THREAD

**Issue by byronmccollum** [ansible/ansible-modules-core#121]
See https://github.com/ansible/ansible/issues/7586

**Comment by benhoyt** [comment]
Yes, this issue still exists in the latest (development) version, as the response content is still always decoded with `unicode_escape`, breaking JSON (and probably other content types too). Copying from the issue over at ansible/ansible for the record:

##### Issue Type:

Bug Report

##### Ansible Version:

ansible 1.5.4

##### Environment:

Ubuntu 12.04

##### Summary:

uri module calls `content.decode('unicode_escape')` on response, which means escape sequences in JSON responses aren't handled properly

##### Steps To Reproduce:

You can test this by running this ansible command line against a test URL I uploaded:

```
ansible localhost -m uri -a "url=https://s3.amazonaws.com/oresource/python_version.json return_content=true"
```

##### Expected Results:

Output should have a "json" key with the PythonVersion string decoded properly.

##### Actual Results:

```
localhost | success >> {
    "accept_ranges": "bytes",
    "changed": false,
    "content": "{\"PythonVersion\": \"2.7.6 (default, May 16 2014, 16:48:38) \n[GCC 4.4.7 20120313 (Red Hat 4.4.7-4)]\"}",
    "content_length": "100",
    "content_location": "https://s3.amazonaws.com/oresource/python_version.json",
    "content_type": "application/json",
    "date": "Wed, 28 May 2014 21:00:47 GMT",
    "etag": "\"2ad455693e539d42df070a367a8f16a3\"",
    "last_modified": "Wed, 28 May 2014 20:52:45 GMT",
    "redirected": false,
    "server": "AmazonS3",
    "status": 200,
    "x_amz_id_2": "Li22a6y8Vb+mYo8AZTN04QolNK+jlJRD8O3nS/rsmUxTpQMAT0ie8K6rbXOf6sDE",
    "x_amz_request_id": "E077ABA55DC7E255"
}
```

##### Further comments

Note that the output does not include the "json" key due to the above bug.

I've been trying to use the "uri" module and hitting an HTTP endpoint which returns some JSON with header `Content-Type: application/json`. However, one of the strings in my JSON has a line feed (`\n` control character) in it. This is correctly escaped, and is valid JSON. However, the "uri" module does the following on the response content (see [here](https://github.com/ansible/ansible/blob/devel/library/network/uri#L307)):

```
unicode(content.decode('unicode_escape'))
```

This decodes the `\n` string to a literal line feed, and causes the `json.loads()` further down to fail with an exception like:

```
ValueError: Invalid control character at: line 1 column 58 (char 58)
```

There's a catch-all exception around the json.loads() that silently ignores this error without setting the `.json` key.

So two things:

1) Why does the uri module do this extra (un)escaping, effectively breaking the content in this case?

2) This issue was hardish to debug as the `json.loads()` has a catch-all except around it. (Catch-all except clauses are bad practice for this reason.) I recommend changing it to a "except ValueError, error" and then call `module.fail_json()` with a string representation of the error in the response.

**Comment by bbradbury** [comment]
Just ran into this in 1.7.2-- it seems as though there's not a sensible way to add a short JSON literal to uri's body= as a result of this bug? Looks like the only work around is to load the json from a file, which is a bummer.

**Comment by benhoyt** [comment]
Anyone from Ansible want to comment on my questions (1 and (2 above?

**Comment by bbradbury** [comment]
Ah ha-- I sorted out JSON literals thusly, using double quotes, single quotes, and escaped double quotes, respectively-- @benhoyt not sure if something like this would help you too?

```
 pre_tasks:
   - uri: "url=http://localhost:9200/_cluster/settings method=PUT
           body='{ \"transient\" : { \"cluster.routing.allocation.enable\" : \"none\" } }'
           status_code=201"
```

**Comment by benhoyt** [comment]
Oh, I see -- no I think yours is a separate issue. This issue (and the one I was having) is about JSON values in a uri _response_.

**Comment by abadger** [comment]
Unfortunately, even though #8027 fixes the escaping issue the code is still badly broken.  It looks like we can't use unicode_escape or raw_unicode_escape here because they unconditionally use latin-1 for the underlying codec.  Since the json specification uses utf-8 unconditionally, there will be a large class of return values that will be broken by using those.

I'm reopening this so that more testing can be performed once I generate a fix (hopefully, later today).

**Comment by abadger** [comment]
This should have been fixed in ansible/ansible-modules-core#537 however there are some other issues with that code. I've just submitted ansible/ansible-modules-core#1331 to try to fix those issues. If anyone is still interested in this, please test that PR to be sure it doesn't cause a regression in this fix.
